grudge/reductions.py

```diff
@@ -159,12 +162,23 @@ def nodal_sum_loc(dcoll: DiscretizationCollection, dd, vec) -> float:
         return sum(nodal_sum_loc(dcoll, dd, vec[idx])
                    for idx in np.ndindex(vec.shape))
 
-    # FIXME: do not force result onto host
     actx = vec.array_context
-    return sum([actx.np.sum(grp_ary) for grp_ary in vec])
+
+    if hasattr(actx, "_force_device_scalars"):
+        force_device_scalars = actx._force_device_scalars
+    else:
+        force_device_scalars = True
+
+    def device_sum(x):
+        result = actx.np.sum(x)
+        if not force_device_scalars:
+            result = actx.from_numpy(np.array(result))
+        return result
```

Comment:
Is this extra work required? In the `not force_device_scalars` case (which is just a deprecated `pyopencl` thing), the reductions always return host scalars, but this forces the grudge reductions to return device scalars, which seems inconsistent.

Does anything break if this just does the `actx.np.sum` and uses whatever that returns? [`to_numpy`](https://github.com/inducer/arraycontext/blob/c5ed317ac595d4ee1e222930d6e0eb0b3315f366/arraycontext/impl/pyopencl/__init__.py#L164-L165) should let these pass through, hopefully.

Reply:
I took another look at this, and I think I can safely remove the `force_device_scalars` handling from `nodal_sum_loc`, but not the others. In `nodal_min_loc`/`nodal_max_loc`, I need device scalar results for each `grp_array` so they can subsequently be passed into `actx.np.minimum`/`actx.np.maximum`. I tried to tweak the return values a little in 214e865 to make them more compatible with `force_device_scalars`. Can you tell me what you think?

Reply:
I'm definitely ok with that, was just wondering what didn't quite work. When we introduced that `force_device_scalars` thing, the hope was that `to_numpy` would just pass it through so that user code won't have to check.

In case you tried this already, what breaks if you just write, e.g.
```
    return reduce(
        lambda acc, grp_ary: actx.np.minimum(acc, actx.np.min(grp_ary)),
        vec, np.inf)
```
?

Reply:
Without the `force_device_scalars` handling I was seeing errors like [these](https://github.com/inducer/grudge/runs/3861787012?check_suite_focus=true) (the `AttributeError: 'numpy.float64' object has no attribute 'queue'`).